rust/template/src/d3main.rs

```diff
@@ -0,0 +1,270 @@
+        let f = fact!(d3_application::Error,
+                      text => text,
+                      line => line,
+                      instance => self.uuid.clone().into_record(),
+                      filename => filename,
+                      functionname => functionname);
+        self.error.clone().send(f);
+    }
+
+    // does it make sense to try to use the HDDLog record evaluation?
+    fn eval(&self, input: Batch) -> Result<Batch, Error> {
+        // would like to implicitly convert batch to ddvalue_batch, but i cant, because i need an
+        // evaluator, and its been deconstructed before we get here...
+        let mut upd = Vec::new();
+        let b = DDValueBatch::from(self, input)?;
+
+        for (relid, v, _) in &b {
+            upd.push(Update::Insert { relid, v });
+        }
+
+        self.h.transaction_start()?;
+        self.h.apply_updates(&mut upd.clone().drain(..))?;
+        Ok(DDValueBatch::from_delta_map(
+            self.h.transaction_commit_dump_changes()?,
+        ))
+    }
+
+    fn id_from_relation_name(&self, s: String) -> Result<usize, Error> {
+        let s: &str = &s;
+        match Relations::try_from(s) {
+            Ok(r) => Ok(r as usize),
+            Err(_) => Err(Error::new(format!("bad relation {}", s))),
+        }
+    }
+
+    fn localize(&self, rel: usize, v: DDValue) -> Option<(Node, usize, DDValue)> {
+        match self.h.d3log_localize_val(rel, v.clone()) {
+            Ok((Some(n), r, v)) => Some((n, r, v)),
+            Ok((None, _, _)) => None,
+            Err(_) => None,
+        }
+    }
+
+    // doesn't belong here. but we'd like a monotonic wallclock
+    // to sequence system events. Also - it would be nice if ddlog
+    // had some basic time functions (format)
+    fn now(&self) -> u64 {
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_millis() as u64
+    }
+
+    fn record_from_ddvalue(&self, d: DDValue) -> Result<Record, Error> {
+        Ok(d.into_record())
+    }
+
+    fn relation_name_from_id(&self, id: usize) -> Result<String, Error> {
+        match relid2name(id) {
+            Some(x) => Ok(x.to_string()),
+            None => Err(Error::new("unknown relation id".to_string())),
+        }
+    }
+
+    // xxx - actually we want to parameterize on format, not on internal representation
+    fn serialize_batch(&self, b: DDValueBatch) -> Result<Vec<u8>, Error> {
+        let w = SerializeBatchWrapper { b };
+        let encoded = serde_json::to_string(&w)?;
+        Ok(encoded.as_bytes().to_vec())
+    }
+
+    fn deserialize_batch(&self, s: Vec<u8>) -> Result<DDValueBatch, Error> {
+        let s = std::str::from_utf8(&s)?;
+        let v: SerializeBatchWrapper = serde_json::from_str(&s)?;
+        Ok(v.b)
+    }
+}
+
+pub fn start_d3log() -> Result<(), Error> {
+    let (uuid, is_parent) = if let Some(uuid) = std::env::var_os("uuid") {
```

Comment:
We discussed that this is a temporary hack.  I propose the following solution for the startup protocol:

1. Ignore initial outputs.  I don't think they are important and I don't think we should rely on them for anything in D3log.  In fact they may disappear from DDlog altogether at some point.

2. `d3_application.dl` should include an input relation (`D3Config`?) that defines its startup configuration.  The runtime populates this relation with a single record in the very first transaction and never touches is again (and the application can rely on this behavior).  The record consists of at least `parent: Option<Node>`.  `parent = None` indicates that we are the root of the local hierarchy. Additionally it can contain arbitrary state passed to it by parent, e.g., `argv: Vec<string>`.

One advantage of this design is that we can do more things, like advertising the Stdout port relationally.